### PR TITLE
[WIP] matrices: trace of MatAdd

### PR DESCRIPTION
**Pull request by lokeshh**
Earlier trace(MatAdd(X, Y)) was failing with type error
'mix of Matrix and Scalar symbols'. This was because
trace(MatAdd(X, Y)) evaluated to MatAdd(trace(X), trace(Y))
and since validate function would reject trace(X) as Matrix
object it was throwing an exception.

It has been fixed by evaluating trace(MatAdd(X, Y)) to
trace(Z) where Z is what we have after evaluating
MatAdd(X, Y).

Fixes issue #9028

### CODE REVIEW

**Comment by cbm755** [comment]
Awesome, thanks for looking into this.

Can you hold off until after #9009?  It has an XFAIL test for this, which this PR should be able to clean up too.

I was going to change it to `return Add(*[Trace(arg).doit() for arg in self.args]).doit()` but your approach seems simpler!  But should we worry about the following?

````
X = Matrix([[1, 2], [3, 4]]);
Y = MatrixSymbol('Y', 2, 2);
q = MatAdd(X, Y, 2*X)
Trace(q).doit()
Out[45]: 
Trace(Matrix([
[3,  6],
[9, 12]]) + Y)
````
In my proposed PR that goes to `15 + Trace(y)`.  Is that better?  Or is `Add` dangerous here somehow?

**Comment by cbm755** [comment]
I guess I don't see the idea of the trace distributing across the addition in your patch.  OTOH, it does fix the issue I was having...  I'm not sure.  What do you think?

**Comment by lokeshh** [comment]
After seeing your way, I would prefer your way of doing things but I think there is still some work to be done. 

For instance,

```python
    X = Matrix([1])
    Y = MatrixSymbol('Y', 1, 1)
    >>> trace(X + Y - Y)
    1                   # This looks good :)
    >>> trace(MatAdd(X, Y, -Y))
    1 - Trace(Y) + Trace(Y)                    # This doesn't look good :(
```

**Comment by lokeshh** [comment]
It seems the problem is with `trace` not `Add`. 

Take a look at this:
```python
>>> Y = MatrixSymbol('Y', 1, 1)
>>> trace(-Y)
-Trace(Y)
>>> -trace(Y)
-Trace(Y)
>>> trace(-Y) == -trace(Y)
False

**Comment by cbm755** [comment]
Well perhaps we don't need to fix everything all at once ;-)  I like the idea of adding an XFAIL test at least for your first `X + Y - Y` example.  You could file an issue, then put the issue number as a comment in the test.  #9009 is merged so this PR could also remove the XFAIL decorator from my test in tests/test_trace

Do you want to do those things?  Let me know if you need help.

Also: take a look at the `srepr` for your second example.  Does this still happen after #9009?  I think doit(deep) would take the MatrixSymbol out of the MatMul, but have not tried yet.

**Comment by cbm755** [comment]
Also, #9009 had fixes for MatAdd so perhaps things look better now :)

**Comment by cbm755** [comment]
By the way: add "Fixes #9028." in the body of one of your commit messages and when this is merged, it will automatically close the bug.

**Comment by lokeshh** [comment]
Very impressive! after applying #9009 everything seems good now. All examples I can think of are passing like:
```python
>>> X = Matrix([2])
>>> Y = MatrixSymbol('Y', 1, 1)
>>> trace(MatAdd(X, Y, -3*Y))
2 - 2*Trace(Y)
```
> Well perhaps we don't need to fix everything all at once ;-) I like the idea of adding an XFAIL test at least for your first X + Y - Y example. You could file an issue, then put the issue number as a comment in the test. #9009 is merged so this PR could also remove the XFAIL decorator from my test in tests/test_trace

I am on it.

> Also: take a look at the `srepr` for your second example.

Where do I look for `srepr`?

**Comment by cbm755** [comment]
> Where do I look for `srepr`?

Its a command that shows you the string representation.  E.g.,

````
U = MatixSymbol('U', 2, 2)
srepr(U)
````

**Comment by lokeshh** [comment]
```python
@XFAIL
def test_Trace_MatAdd_doit():
    # FIXME: Issue #9028.
    X = Matrix([[1, 2], [3, 4]])
    q = MatAdd(X, 2*X)
    assert Trace(q).doit(deep=False).arg == q
```

I couldn't pass this test since Trace gets distributed when applied on MatAdd.

There's another problem which I am trying to resolve and that is:
```python
trace(Y) + trace(-Y)
```
is returning
```python
-Trace(Y) + Trace(Y)
```
BTW `srepr` is proving to be very useful. Thanks for that!

**Comment by cbm755** [comment]
1.  I will think about my test...

2.  I think that (uppercase) `Trace(Y) + Trace(-Y)` is supposed to stay unevaluated: but I think it could go to `Trace(Y) - Trace(Y)`.  (lowercase trace is a different matter).  Both should be tested.

3.  Suppose `X` is a concrete ImmutableMatrix (like the 2x2 in my test).  Then `Trace(X) + Trace(-X)` should also not evaluate.  With `.doit()` it should go to zero.  I'm not sure about `.doit(deep=False)`, probably should canonicalize but not evaluate (whatever that means in this context!)   Actually `Trace(X) + Trace(-X)` looks like another bug to me!

4.  Perhaps we are trying to do too much here, too quickly.  Are we mixing up two different issues?  Possible to just fix MatAdd?  Or are these too interconnected?

File: sympy/matrices/expressions/matadd.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def _eval_adjoint(self):
 
     def _eval_trace(self):
         from trace import Trace
-        return MatAdd(*[Trace(arg) for arg in self.args]).doit()
+        return Add(*[Trace(arg).doit() for arg in self.args])
```
Comment (cbm755) [inline sympy/matrices/expressions/matadd.py]: I think I stand by my XFAIL test.

Perhaps we should drop this `.doit()`?  But will something else fail?

Comment (cbm755) [inline sympy/matrices/expressions/matadd.py]: Wait, maybe I stand by it not evaluating... but test might be wrong in current form.  Will think about it more...

**Comment by lokeshh** [comment]
Well, my main focus was to write tests for MatAdd but I got confused what tests should I write. Then I looked at other Classes like Adjoint, Transpose to take inspiration but came to know Trace is different from them in lot of aspects. I wished for a uniformity among them all at how they evaluate and strongly felt that Trace should be changed. For example,

```python
>>> adjoint(Y) + adjoint(-Y)
0
```
while Trace works as follows
```python
>>> trace(Y) + trace(-Y)
-Trace(Y) + Trace(Y)
```

I think I got too carried away with them and lost the main focus which was to rectify MatAdd. I think we have to refrain from writing any more tests unless we make sure what are we looking for. So, I guess I will leave it like that for now and do things one at a time. 

What do you say?

File: sympy/matrices/expressions/matadd.py (hunk 1 above)
Comment (lokeshh) [inline sympy/matrices/expressions/matadd.py]: I am dropping the `.doit()`. It's better this way.

Comment (cbm755) [inline sympy/matrices/expressions/matadd.py]: Will lowercase trace() function properly without it?

**Comment by cbm755** [comment]
Perhaps we can file bugs for these other issues (and also search for existing issues).   (I think I agree in principle with the Trace/trace changes, or at least with your examples!).

**Comment by cbm755** [comment]
Maybe this the test I want?  One must be very careful changing tests...  ImmutableMatrix is b/c of #9043.

````
def test_Trace_MatAdd_doit():
    X = ImmutableMatrix([[1, 2], [3, 4]])
    q = MatAdd(X, 2*X)
    assert Trace(q).doit(deep=False) == Trace(X) + Trace(2*X)
````

**Comment by lokeshh** [comment]
Cool! I can pass this test.

File: sympy/matrices/expressions/matadd.py (hunk 1 above)
Comment (lokeshh) [inline sympy/matrices/expressions/matadd.py]: Yep

File: sympy/matrices/expressions/tests/test_trace.py [hunk 2]
```diff
@@ -53,12 +53,10 @@ def test_Trace_doit():
...
 
 
-@XFAIL
 def test_Trace_MatAdd_doit():
-    # FIXME: Issue #9028.
-    X = Matrix([[1, 2], [3, 4]])
+    X = ImmutableMatrix([[1, 2], [3, 4]])
     q = MatAdd(X, 2*X)
-    assert Trace(q).doit(deep=False).arg == q
+    assert Trace(q).doit(deep=False) == Trace(X) + Trace(2*X)
```
Comment (cbm755) [inline sympy/matrices/expressions/tests/test_trace.py]: might as well test that `Trace(q).doit()` gives the correct answer here too.

**Comment by cbm755** [comment]
In general, you shouldn't push without checking the local test works: just do `test('sympy/matrices/expressions/tests/test_trace.py')`.  (of course, some other unexpected test can always fail).

(you need to add ImmutableMatrix to the imports at the top.)

**Comment by lokeshh** [comment]
> In general, you shouldn't push without checking the local test works: just do test('sympy/matrices/expressions/tests/test_trace.py'). (of course, some other unexpected test can always fail).

Oops, I forgot to run the tests! it won't happen again.
BTW where am I supposed to run `test(sympy/.../test_trace.py)`?

> (you need to add ImmutableMatrix to the imports at the top.)

Thanks for mentioning that out :+1:

File: sympy/matrices/expressions/tests/test_trace.py (hunk 2 above)
Comment (lokeshh) [inline sympy/matrices/expressions/tests/test_trace.py]: Done!

**Comment by lokeshh** [comment]
So what are we left with now? These are I can think of:

a). Deciding what should `Trace(Y) + Trace(-Y)` and `trace(Y) + trace(-Y)` evaluate to?

b). What should `.doit(deep = False)` supposed to do? 
```python
>>> Adjoint(MatAdd(X, 2*X)).doit(deep = False)
```
gives
```python
3*Adjoint(Y)
```
while
```python
>>> Trace(MatAdd(X, 2*X)).doit(deep = False)
```

gives

```python
Trace(2*Y) + Trace(Y)
```
(I am with changing `Adjoint` and other functions to match the `Trace.doit(deep=False)` situation which seems more intuitive. What do you say on this?)


c). #9043

(What else? Will add more to this list as they come up)

**Comment by lokeshh** [comment]
I have no idea why the test is failing. It's working fine in my system.

**Comment by cbm755** [comment]
failing test: something different in python 3 apparently, not sure either.  Install on Python 3 and play with it?

I don't see a test yet for this: https://github.com/sympy/sympy/pull/9031#issuecomment-74974127

If I were working on this, I would go read up on what deep=False is supposed to do in other cases (away from MatrixExpr).

**Comment by lokeshh** [comment]
Yep, its failing with Python3 as your suggested. `trace.Trace` object is passed to sympify instead of `sympy.matrices.expressions.trace.Trace`. Any idea why?

Where can I read about function of deep = False?

**Comment by cbm755** [comment]
1.  No idea :(

2.  maybe grep source code for `deep`?

**Comment by lokeshh** [comment]
Much of the code still uses old-style classes where as support of old-style classes has been removed from Python3. I don't know if that's what is causing the bug but shouldn't the whole code need to be revisited to replace the old-style classes with new style classes?

**Comment by lokeshh** [comment]
Hahaha... you won't believe what happened!!!

I figured out the bug. It because of this:
`.`

Phew... I read whole bunch of pages about differences in Python2 and Python3 and in the end the problem was just a dot.

**Comment by cbm755** [comment]
Heh :)

I think you should drop this MatMul change: its incorrect as is, b/c `tr(A*B)` is not `tr(A)*tr(B)` in general.  We should probably add a test to protect this property.

I've thought about `Trace.doit()`:

1.  I the only way I can see to make deep=False work is to add a deep vs shallow kwarg to `._eval_trace` or add some kind of `._shallow_trace`.  I do not think we should do that here in this PR.  Maybe we just drop `deep=False` support altogether in `Trace.doit()` but anyway not for this PR.
2.  So my feeling is we leave an XFAIL test for Trace doit `deep=False`.
3.  We should use lowercase `trace` in your patch to `MatAdd._eval_trace()`.  I don't think `trace()` cmd can work correctly otherwise.

I'll think about some tests too.

**Comment by lokeshh** [comment]
Hmm... that was a very big mistake. I should be more careful. Thanks for mentioning that out!

**Comment by cbm755** [comment]
I sent you a PR against your PR :)

File: sympy/matrices/expressions/tests/test_trace.py [hunk 3]
```diff
@@ -44,21 +44,55 @@ def test_Trace():
...
+    assert trace(MatMul(2, X)) == 10
+
+
+def test_Trace_MatAdd_doit():
+    X = ImmutableMatrix([[1, 2, 3]]*3)
+    Y = MatrixSymbol('Y', 3, 3)
+    q = MatAdd(X, 2*X, Y, -3*Y)
+    r = Trace(q)
+    assert r.arg == q
+    assert r.doit() == 18 - 2 * Trace(Y)
```
Comment (cbm755) [inline sympy/matrices/expressions/tests/test_trace.py]: `2*Trace(Y)`, no spaces.

**Comment by cbm755** [comment]
Ok I'm +1 for this.  (you've mentioned various other potential changes: they could be filed elsewhere).

I think it will need a "rebase -i" to squash down to one or two commits.  See [dev guide](https://github.com/sympy/sympy/wiki/Development-workflow).

**Comment by kevinventullo** [comment]
This appears to fail its own test? This test in particular:

```q = MatAdd(X, 2*X)```
```assert Trace(q).doit(deep=False).arg == q```

Even when `deep=False`, Trace will attempt to let its argument evaluate its own trace via `arg._eval_trace()`. MatAdd knows how to take its own trace so you end up getting a number if X is explicit.

**Comment by cbm755** [comment]
@kevinventullo isn't that test XFAIL'd?  My understanding is @lokeshh set out to solve #9028, which this does.

**Comment by cbm755** [comment]
@lokeshh are you able to rebase and squash these commits?

**Comment by lokeshh** [comment]
Nope, I have got things messed up.

When I did `git rebase --interactive 28fd02e`, there were commits from other branches to. That was probably because I frequently kept synching with master. Anyway I squashed those commits, some of them weren't able to because of some thing I couldn't understand. I pushed them anyhow and ended up many push so many commits now.

**Comment by cbm755** [comment]
Ok @lokeshh I did the squash and filed #9063.  (AFAICT, its still attributed to you.)

**Comment by lokeshh** [comment]
Wow... You always have the rescue plan! I was worrying what's going to happen now. 

Thank you very much for helping me so many times.

I am looking forward to fix #9043 now.

**Comment by cbm755** [comment]
We fixed this (using @lokeshh's work) in #9063.

**Pull request closed without merging**